sympy/polys/matrices/sdm.py

```diff
@@ -970,3 +1012,35 @@ def sdm_particular_from_rref(A, ncols, pivots):
         if Ain is not None:
             P[j] = Ain / A[i][j]
     return P
+
+
+def sdm_applyfunc(M, shape, f, domain):
+    nrows, ncols = shape
+    Mf = {}
+    for i in range(nrows):
+        Mf[i] = {}
+        for j in range(ncols):
+            try:
+                domain_elem = domain.convert(f(M[i][j]))
+                if not domain.is_zero(domain_elem):
+                    Mf[i][j] = domain_elem
+            except KeyError:
+                domain_elem = domain.convert(f(domain.zero))
```

Comment:
Here `f(domain.zero)` should only be called once. If it gives zero then `sdm_applyfunc_nonzero` can be used. Otherwise `sdm_applyfunc` should effeciently initialise a matrix full of the result and then only loop over the nonzero elements after.